### ISSUE TITLE
Scalatra decodes splat parameter 'too much'

### ISSUE THREAD

**Issue by rvanheest** [scalatra/scalatra#969]
Let's say I have a servlet route defined with Scalatra using a `splat` argument containing a path to a file:

```scala
get(":foo/*") {
	val foo = params("foo")
	val tail = multiParams("splat")

	println(foo)
	println(tail)
}
```

I have a file that I want to access via this servlet, with path `data/my%2Ffile.txt`. Please note, the filename of this file is `my%2Ffile.txt`!!!

In my client I apply character encoding using [Guava's `PercentEscapter`](https://guava.dev/releases/18.0/api/docs/com/google/common/net/PercentEscaper.html), which amongst others escapes `%` with `%25`. The client does a `GET` on `<base url>/foobar/data/my%252Ffile.txt`.

In the above Scalatra route, the `val foo` will be `foobar` and `val tail` is expected to be `data/my%2Ffile.txt` (since it should decode `%25` to `%` again). However, the result is not this expected value, but rather `data/my/file.txt`.

Using the debugger I found that [`ScalatraServlet.requestPath`](https://github.com/scalatra/scalatra/blob/9f0a2965a815be9a0cbbefb1bdc4a4a4272bc586/core/src/main/scala/org/scalatra/ScalatraServlet.scala#L34-L43) decodes `data/my%252Ffile.txt` to `data/my%2Ffile.txt` in `UriDecoder.firstStep`, but afterwards it further decodes `data/my%2Ffile.txt` to `data/my/file.txt` in [`ScalatraBase.setMultiparams`](https://github.com/scalatra/scalatra/blob/9f0a2965a815be9a0cbbefb1bdc4a4a4272bc586/core/src/main/scala/org/scalatra/ScalatraBase.scala#L359-L367) via `UriDecoder.secondStep`, which literally replaces `%2F` with `/`. This second step is not what I want to have happened to my servlet route.

Is there any reason why this second step is performed? Does anybody know how to avoid this behavior or work around it?

**Comment by magnolia-k** [comment]
It is below commit that has been incorporated twice, but the reason is not written

https://github.com/scalatra/scalatra/commit/8b310b9eb9edadc4cc93577e9466cb3a8985f797

In general, decoding twice is not appropriate. I think it's better to fix it so that it doesn't decode twice.

**Comment by rvanheest** [comment]
@magnolia-k nice digging in fairly old code!!! Thanks a lot.
For reference, this commit is part of https://github.com/scalatra/scalatra/pull/189, which fixed https://github.com/scalatra/scalatra/issues/52.

I assume you propose the following change (haven't tested this yet, will do later this week):

```diff
- val routeParams = matchedRoute.map(_.multiParams).getOrElse(Map.empty).map { case (key, values) =>
-   key -> values.map(UriDecoder.secondStep(_))
- }
+ val routeParams = matchedRoute.map(_.multiParams).getOrElse(Map.empty)
```

as well as the naming convention of `firstStep` and `secondStep`?

> I think it's better to fix it so that it doesn't decode twice.

I tend to agree with that conclusion. However, since it has been so long in the library, I worry other applications explicitly or implicitly depend on this two-step process. Any clues/suggestions?

**Comment by magnolia-k** [comment]
I investigated other language frameworks (Pers Kossy, Ruby Sinatra).

 - Route matching is done against PATH_INFO.
 - PATH_INFO is decoded according to the CGI standard
   - Ruby's Rack doesn't decode PATH_INFO, but Sinatra decodes and matches.
   - Perl's Plack passes the decoded PATH_INFO to the framework.

In WAF, if route matching is performed using decoded PATH_INFO,% 2F cannot be distinguished from '/', but it seems to be generally not a problem.
(This is a proven method for other language frameworks.)

In addition, the current Scalatra path handling seems to have the following problems:

 - In matrix URI (a format in which parameters are added with a semicolon), only the pattern in which a semicolon is added at the end of the path is considered, but the matrix URI can be assigned to any hierarchy. When added to any hierarchy, Scalatra deletes everything after the semicolon, so the correct path cannot be generated.According to the Servlet standard, the PATH_INFO is passed to the framework after the matrix URI element is properly deleted.
 - Since parameters are not deleted, a route with parameters can be defined.This is a very different behavior from other frameworks and is not a function that should be provided by the framework from the standpoint of maintainability.

Perhaps the current Scalatra code wanted to distinguish between '/' and% 2F in the path. However, this requires complicated decoding. In other frameworks, the process around the path is kept simple by choosing not to distinguish between '/' and% 2F in the path.

By not distinguishing '/' and% 2F in the path, the code becomes simple, and considering that other edge cases are handled correctly, I think that Scalatra should also switch to route matching using PATH_INFO It is.

However, this change introduces many incompatibilities. Even if the `secondStep` method is abolished and only two-step decoding is stopped, it becomes incompatible.

As a result of rewriting to PATH_INFO base as a trial, it was necessary to rewrite a large number of test codes)

A fix in Scalatra 2.8 or later seems appropriate.

**Comment by takezoe** [comment]
@magnolia-k Thanks for detailed investigation. Scalatra 2.7 isn't GA yet and Scalatra 2.8 won't come within a few years at least. I wonder if we can fix this in Scalatra 2.7.

**Comment by magnolia-k** [comment]
https://github.com/scalatra/scalatra/pull/985
